File: Node/src/dto/objects/Complex.ts
import { ValidateNested } from "class-validator"
import { Price } from "../primitives/Price"
import { Int } from "../tsPrimitives/Int"
import { UUID } from "../tsPrimitives/UUID"
import { Procedure } from "./Procedure"

export namespace Complex {

    export namespace Parameters {
        
        export class Retrieve {
            @ValidateNested()
            salons?: UUID[]

            @ValidateNested()
            employees?: UUID[]

            constructor(
                salons?: UUID[], 
                employees?: UUID[]
            ) {
                this.salons = salons
                this.employees = employees
            }
        }

        export class Create {
            description?: string
            alias?: string

            @ValidateNested()
            price: Price

            @ValidateNested()
            duration: Int

            @ValidateNested()
            procedureIds: UUID[]

            constructor(
                price: Price,
                duration: Int,
                description?: string,
                alias?: string,
                procedureIds: UUID[] = []
            ) {
                this.price = price
                this.duration = duration
                this.description = description
                this.alias = alias
                this.procedureIds = procedureIds
            }
        }

        export class Patch {
            description?: string
            alias?: string

            @ValidateNested()
            price: Price

            @ValidateNested()
            duration: Int

            @ValidateNested()
            procedureIds: UUID[]

            constructor(
                price: Price,
                duration: Int,
                description?: string,
                alias?: string,
                procedureIds: UUID[] = []
            ) {
                this.price = price
                this.duration = duration
                this.description = description
                this.alias = alias
                this.procedureIds = procedureIds
            }
        }
    }

    export namespace Responses {
        
        export class Full {
            description?: string
            alias?: string

            @ValidateNested()
            id: UUID

            @ValidateNested()
            price: Price

            @ValidateNested()
            duration: Int
            
            @ValidateNested()
            procedures: Procedure.Responses.Partial[]

            constructor(
                id: UUID,
                price: Price,
                duration: Int,
                description?: string,
                alias?: string,
                procedures: Procedure.Responses.Partial[] = []
            ) {
                this.id = id
                this.price = price
                this.duration = duration
                this.description = description
                this.alias = alias
                this.procedures = procedures
            }
        }

        export class Partial {
            description?: string
            alias?: string

            @ValidateNested()
            id: UUID
            
            @ValidateNested()
            price: Price

            @ValidateNested()
            duration: Int
            
            @ValidateNested()
            procedures: Procedure.Responses.Partial[]

            constructor(
                id: UUID,
                price: Price,
                duration: Int,
                procedures: Procedure.Responses.Partial[],
                description?: string,
                alias?: string,
            ) {
                this.id = id
                this.price = price
                this.duration = duration
                this.description = description
                this.alias = alias
                this.procedures = procedures
            }
        }
    }
}

File: Android/shared/src/main/kotlin/shared/dto/objects/Complex.kt
@file:UseSerializers(UUIDSerializer::class)

package shared.dto.objects

import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers
import shared.dto.primitives.Price
import shared.dto.protocols.Parametable
import shared.dto.protocols.Responsable
import shared.serializers.UUIDSerializer
import java.util.*

object Complex {
    data object Parameters {
        @Serializable
        data class Retrieve(
            val salons: List<UUID>? = null,
            val employees: List<UUID>? = null,
        ) : Parametable()

        @Serializable
        data class Create(
            val price: Price,
            val duration: Int,
            val description: String?,
            val alias: String?,
            val procedureIds: List<UUID>,
        ) : Parametable()

        @Serializable
        data class Patch(
            val price: Price,
            val duration: Int,
            val description: String?,
            val alias: String?,
            val procedureIds: List<UUID>,
        ) : Parametable()
    }

    data object Responses {
        @Serializable
        data class Full(
            val id: UUID,
            val price: Price,
            val duration: Int,
            val description: String? = null,
            val alias: String? = null,
            val procedures: List<Procedure.Responses.Partial>
        ) : Responsable

        @Serializable
        data class Partial(
            val id: UUID,
            val price: Price,
            val duration: Int,
            val description: String? = null,
            val alias: String?,
            val procedures: List<Procedure.Responses.Partial>
        ) : Responsable
    }
}
